OutOfSchool/OutOfSchool.WebApi/Services/Gcp/GcpImagesStorageSynchronizationService.cs

```diff
@@ -0,0 +1,138 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Apis.Storage.v1.Data;
+using Google.Cloud.Storage.V1;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+using OutOfSchool.Services.Models;
+using OutOfSchool.Services.Models.Images;
+using OutOfSchool.Services.Repository;
+using OutOfSchool.Services.Repository.Files;
+using OutOfSchool.WebApi.Models;
+using Object = Google.Apis.Storage.v1.Data.Object;
+
+namespace OutOfSchool.WebApi.Services.Gcp
+{
+    /// <summary>
+    /// Represents a gcp image files sync storage. It's used to synchronize gcp files with a database.
+    /// </summary>
+    public class GcpImagesStorageSynchronizationService : IGcpStorageSynchronizationService
+    {
+        private const string ListObjectOptionsFields = "items(name,timeCreated),nextPageToken";
+
+        // TODO: must be changed to approximately 500 after testing
+        private const int ListObjectOptionsPageSize = 50;
+
+        private readonly ILogger<GcpImagesStorageSynchronizationService> logger;
+        private readonly IImageFilesStorage imagesStorage;
+        private readonly IGcpImagesSyncDataRepository gcpImagesSyncDataRepository;
+
+        public GcpImagesStorageSynchronizationService(
+            ILogger<GcpImagesStorageSynchronizationService> logger,
+            IImageFilesStorage imagesStorage,
+            IGcpImagesSyncDataRepository gcpImagesSyncDataRepository)
+        {
+            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            this.imagesStorage = imagesStorage ?? throw new ArgumentNullException(nameof(imagesStorage));
+            this.gcpImagesSyncDataRepository = gcpImagesSyncDataRepository ?? throw new ArgumentNullException(nameof(gcpImagesSyncDataRepository));
+
+            GetAllSyncFunctions = new List<Func<IEnumerable<string>, Task<List<string>>>>
+            {
+                // Entity images
+                gcpImagesSyncDataRepository.GetIntersectWorkshopImagesIds,
+                gcpImagesSyncDataRepository.GetIntersectProviderImagesIds,
+
+                // Entity cover images
+                gcpImagesSyncDataRepository.GetIntersectWorkshopCoverImagesIds,
+                gcpImagesSyncDataRepository.GetIntersectTeacherCoverImagesIds,
+                gcpImagesSyncDataRepository.GetIntersectProviderCoverImagesIds,
+            };
+        }
+
+        private List<Func<IEnumerable<string>, Task<List<string>>>> GetAllSyncFunctions { get; }
+
+        /// <inheritdoc/>
+        public async Task SynchronizeAsync(CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                logger.LogDebug("Gcp storage synchronization was started");
+
+                // gcp returns data by local time
+                var dateTime = DateTime.Now.AddMinutes(-1);
```

Comment:
maybe skip more then a minute (e.g., 1 hour or even 1 day) to be sure we won't delete files that are in process of writing to storage&database (if this job might take long time to complete)